File: RateLimiter2/TokenBucket.js
const RateLimiterStrategy = require('./RateLimiterStrategy');

class TokenBucket extends RateLimiterStrategy {
  constructor(capacity, refillRate) {
    super();
    this.capacity = capacity;
    this.refillRate = refillRate;
    this.bucket = new Map();
  }

  refill(ip) {
    const currentTime = Date.now();
    const ipRequests = this.bucket.get(ip);

    const timeElapsed = (currentTime - ipRequests.lastRefilledAt) / 1000;
    const refills = Math.floor(timeElapsed * this.refillRate);

    if (refills > 0) {
      this.bucket.set(ip, {
        tokens: Math.min(this.capacity, ipRequests.tokens + refills),
        lastRefilledAt: currentTime,
      });
    }
  }

  isAllowed(ip) {
    if (!this.capacity) return false;

    if (!this.bucket.has(ip)) {
      this.bucket.set(ip, {
        tokens: this.capacity - 1,
        lastRefilledAt: Date.now(),
      });

      return true;
    }

    if (this.refillRate) {
      this.refill(ip);
    }

    const ipRequests = this.bucket.get(ip);
    if (ipRequests.tokens > 0) {
      this.bucket.set(ip, {
        tokens: ipRequests.tokens - 1,
        lastRefilledAt: ipRequests.lastRefilledAt,
      });
      return true;
    }

    return false;
  }
}

module.exports = TokenBucket;

File: Go/RateLimiter2-go/token_bucket.go
package main

import (
	"sync"
	"time"
)

type tokenBucketState struct {
	Tokens         int
	LastRefilledAt int64
}

type TokenBucketStrategy struct {
	mu         sync.Mutex
	Capacity   int
	RefillRate int
	Bucket     map[string]tokenBucketState
}

func NewTokenBucketStrategy(capacity, refillRate int) *TokenBucketStrategy {
	return &TokenBucketStrategy{
		Capacity:   capacity,
		RefillRate: refillRate,
		Bucket:     make(map[string]tokenBucketState),
	}
}

func (t *TokenBucketStrategy) refill(ip string) {
	currentTime := time.Now().UnixMilli()
	ipRequests := t.Bucket[ip]
	timeElapsed := float64(currentTime-ipRequests.LastRefilledAt) / 1000.0
	refills := int(timeElapsed * float64(t.RefillRate))
	if refills > 0 {
		t.Bucket[ip] = tokenBucketState{
			Tokens:         min(t.Capacity, ipRequests.Tokens+refills),
			LastRefilledAt: currentTime,
		}
	}
}

func (t *TokenBucketStrategy) IsAllowed(ip string) bool {
	t.mu.Lock()
	defer t.mu.Unlock()
	if t.Capacity == 0 {
		return false
	}
	if _, ok := t.Bucket[ip]; !ok {
		t.Bucket[ip] = tokenBucketState{
			Tokens:         t.Capacity - 1,
			LastRefilledAt: time.Now().UnixMilli(),
		}
		return true
	}
	if t.RefillRate > 0 {
		t.refill(ip)
	}
	ipRequests := t.Bucket[ip]
	if ipRequests.Tokens > 0 {
		t.Bucket[ip] = tokenBucketState{
			Tokens:         ipRequests.Tokens - 1,
			LastRefilledAt: ipRequests.LastRefilledAt,
		}
		return true
	}
	return false
}
